### ISSUE TITLE
Allow to run two (or more) DVC steps that have shared dependencies at the same time 

### ISSUE THREAD

**Issue by IamGianluca** [iterative/dvc#5454]
Hi,

Thank you so much for the library, I'm a huge fan as it is really game-changing for an effective ML workflow!

Right now, when trying to execute a DVC step that has one or more dependencies shared with another DVC step currently running, we are notified that a lock is in place. This, IMHO, is unnecessary as no DVC step _should_ change an input file ― at least, this should be best practice.

The use case here is that sometimes, we might have enough resources to execute at the same time multiple steps. For instance, if we have multiple GPUs, we might be able to train multiple models using the same input data in parallel. This is actually a workflow most people adopt when having at their disposal multiple GPUs.

I would recommend we drop that lock, and maybe raise an error if a user defines a file both as an input and output file to a DVC step ― maybe this is already done, I haven't had time to check yet.

**Comment by skshetry** [comment]
We had a similar discussion about this on https://github.com/iterative/dvc/issues/4979. It's there to protect users, but I think we do want to make this a default behaviour.
